simulation/interfaces/sim_ui.py

```diff
@@ -0,0 +1,397 @@
+
+"""Simulator web UI using streamlit."""
+
+import os.path
+import sys
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+
+from concurrent.futures import ProcessPoolExecutor
+from io import StringIO
+from typing import Union
+
+import numpy as np
+import pandas as pd
+import streamlit as st
+import yaml
+from core.create_index import create_stream_index
+from core.main import simulate
+from core.shuffle_quality import analyze_shuffle_quality_entropy
+from core.sim_dataset import SimulationDataset
+from core.sim_time import Time
+from core.utils import get_total_batches
+from core.yaml_processing import create_simulation_dataset, ingest_yaml
+from interfaces.interface_utils import get_train_dataset_params
+from interfaces.widgets import (display_shuffle_quality_graph, display_simulation_stats,
+                                get_line_chart, param_inputs)
+
+from streaming.base.util import bytes_to_int, number_abbrev_to_int
+
+# set up page
+st.set_page_config(layout='wide')
+col1, space, col2 = st.columns((10, 1, 6))
+col2.title('Streaming Simulator')
+col2.write('Enter run parameters in the left panel.')
+col2.text('')
+progress_bar = col1.progress(0)
+status_text = col1.empty()
+col1.text('')
+throughput_plot = col2.empty()
+network_plot = col2.empty()
+sim_stats = col2.empty()
+col2.text('')
+shuffle_quality_plot = col2.empty()
+throughput_window = 10
+shuffle_quality_algos = ['naive', 'py1b', 'py1br', 'py1e', 'py1s', 'py2s', 'none']
+
+
+def submit_jobs(shuffle_quality: bool, dataset: SimulationDataset, time_per_sample: float,
+                node_internet_bandwidth: Union[int, str], max_duration: Time):
+    """Submit jobs to the executor for simulation.
+
+    Args:
+        shuffle_quality (bool): Whether to run shuffle quality analysis.
+        dataset (SimulationDataset): Dataset to use for simulation.
+        time_per_sample (float): Time for one device to process one sample.
+        node_internet_bandwidth (Union[int,str]): Internet bandwidth per node.
+        max_duration (Time): Maximum duration of simulation.
+    """
+    total_batches = get_total_batches(dataset=dataset, max_duration=max_duration)
+    node_internet_bandwidth = bytes_to_int(node_internet_bandwidth)
+    cache_limit = dataset.get_cache_limit()
+    gen_sim = simulate(dataset,
+                       time_per_sample,
+                       node_internet_bandwidth,
+                       max_duration,
+                       generator=True)
+    gen_step_times = []
+    gen_step_downloads = []
+    rolling_throughput_data = []
+    immediate_throughput_data = []
+    network_data = []
+    steps = []
+    time_to_first_batch = 0
+    futures = []
+    shuffle_quality_graphed = False
+    # Initialize min_cache_limit to be 0. Will be replaced by the simulated value.
+    min_cache_limit = 0
+    # Define partial function to pass to executor map for simulation.
+    with ProcessPoolExecutor(max_workers=8) as executor:
```

Comment:
nit: Do we need `max_workers=8`? I think it is by default set as a number of logical CPU cores in your system.

Reply:
ooh nice. checked and confirmed you're right!